File: vsletterboxc/vsletterbox.cpp
// vapoursynth-letterbox
// Copyright (c) Akatsumekusa and contributors

// ---------------------------------------------------------------------
// Permission is hereby granted, free of charge, to any person obtaining
// a copy of this software and associated documentation files (the
// "Software"), to deal in the Software without restriction, including
// without limitation the rights to use, copy, modify, merge, publish,
// distribute, sublicense, and/or sell copies of the Software, and to
// permit persons to whom the Software is furnished to do so, subject to
// the following conditions:
// 
// The above copyright notice and this permission notice shall be
// included in all copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
// BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
// ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
// CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
// ---------------------------------------------------------------------

#include <algorithm>
#include <limits>
#include <cmath>
#include <memory>
#include <optional>
#include <cstdint>
#include <cstring>
#include <type_traits>
#include <VapourSynth4.h>
#include <VSHelper4.h>

struct FindData {
    VSNode *clip;
    double  thr;
    VSNode *ref;
    double  ref_thr;
};

template <typename T>
static inline constexpr T get_max_value() {
    if constexpr (std::is_integral_v<T>)
        return std::numeric_limits<T>::max();
    else
        return 1.0;
}
template <typename T>
static inline constexpr T get_min_value() {
    if constexpr (std::is_integral_v<T>)
        return std::numeric_limits<T>::min();
    else
        return 0.0;
}

template <typename T>
static inline double calc_mean(const T * VS_RESTRICT srcp, int width) {
    constexpr T max_value = get_max_value<T>();
    constexpr T min_value = get_min_value<T>();

    std::conditional_t<std::is_same_v<T, uint8_t>,
                       uint32_t,
                       std::conditional_t<std::is_same_v<T, uint16_t>,
                                          uint64_t,
                                          double>>
        sum = 0;
    if constexpr (std::is_integral_v<T>) {
        #pragma clang loop vectorize(assume_safety) interleave(enable)
        for (int x = 0; x < width; x++) {
            #pragma clang fp reassociate(on)
            sum += srcp[x];
        }
    }
    else {
        #pragma clang loop vectorize(assume_safety) interleave(enable)
        for (int x = 0; x < width; x++) {
            #pragma clang fp reassociate(on)
            sum += std::clamp(srcp[x], min_value, max_value);
        }
    }

    return static_cast<double>(sum) / width / max_value;
}

// template <typename T>
// static inline double calc_root_mean_square(const T * VS_RESTRICT srcp, int width) {
//     constexpr T max_value = get_max_value<T>();
//     constexpr T min_value = get_min_value<T>();

//     double sum = 0.0;
//     if constexpr (std::is_integral_v<T>) {
//         #pragma clang loop vectorize(assume_safety) interleave(enable)
//         for (int x = 0; x < width; x++) {
//             #pragma clang fp reassociate(on)
//             const auto x_ = static_cast<double>(srcp[x]);
//             sum += x_ * x_;
//         }
//     }
//     else {
//         #pragma clang loop vectorize(assume_safety) interleave(enable)
//         for (int x = 0; x < width; x++) {
//             #pragma clang fp reassociate(on)
//             const auto x_ = static_cast<double>(std::clamp(srcp[x], min_value, max_value));
//             sum += x_ * x_;
//         }
//     }

//     return std::sqrt(sum / width) / max_value;
// }

template <typename T>
static inline double calc_15_power_mean(const T * VS_RESTRICT srcp, int width) {
    constexpr T max_value = get_max_value<T>();
    constexpr T min_value = get_min_value<T>();

    double sum = 0.0;
    if constexpr (std::is_integral_v<T>) {
        #pragma clang loop vectorize(assume_safety) interleave(enable)
        for (int x = 0; x < width; x++) {
            #pragma clang fp reassociate(on)
            const auto x_ = static_cast<double>(srcp[x]);
            sum += x_ * __builtin_sqrt(x_);
        }
    }
    else {
        #pragma clang loop vectorize(assume_safety) interleave(enable)
        for (int x = 0; x < width; x++) {
            #pragma clang fp reassociate(on)
            __builtin_assume(!isnan(srcp[x]));
            const auto x_ = static_cast<double>(std::clamp(srcp[x], min_value, max_value));
            sum += x_ * __builtin_sqrt(x_);
        }
    }

    return std::pow(sum / width, 2.0 / 3) / max_value;
}

// Incremental calculation of weighted mean and variance, Tony Finch, 2009
template <double alpha = 0.05>
class ExponentiallyWeightedStats {
    static_assert(alpha > 0.0 && alpha < 1.0);
    static constexpr double one_minus_alpha = 1 - alpha;

    bool   init = false;
    double _mean;
    double _var;

public:
    void add_data(double data) {
        if (!init) {
            init  = true;
            _mean = data;
            _var  = 0.0;
        }
        else {
            const double diff =  data - _mean;
            const double incr =  alpha * diff;
            _mean             += incr;
            _var              =  one_minus_alpha * (_var + diff * incr);
        }
    }
    std::optional<double> mean() {
        if (!init)
            return std::nullopt;
        else
            return _mean;
    }
    std::optional<double> stddev() {
        if (!init)
            return std::nullopt;
        else
            return std::sqrt(_var);
    }
};

template <typename T>
static const VSFrame * VS_CC letterbox_search_get_frame(int n, int activationReason, void *instanceData, void **frameData, VSFrameContext *frameCtx, VSCore *core, const VSAPI *vsapi) {
    auto *d = static_cast<FindData *>(instanceData);

    if (activationReason == arInitial) {
        vsapi->requestFrameFilter(n, d->clip, frameCtx);
        vsapi->requestFrameFilter(n, d->ref, frameCtx);
    }
    else if (activationReason == arAllFramesReady) {
        const auto clip = vsapi->getFrameFilter(n, d->clip, frameCtx);
        if (!clip) {
            vsapi->setFilterError("vsletterbox: Failed to get frame from VapourSynth", frameCtx);
            return nullptr;
        }
        const auto ref  = vsapi->getFrameFilter(n, d->ref, frameCtx);
        if (!ref) {
            vsapi->freeFrame(clip);
            vsapi->setFilterError("vsletterbox: Failed to get frame from VapourSynth", frameCtx);
            return nullptr;
        }

        const int height = vsapi->getFrameHeight(clip, 0);
        const int width  = vsapi->getFrameWidth(clip, 0);
        if (height != vsapi->getFrameHeight(ref, 0) ||
            width != vsapi->getFrameWidth(ref, 0)) {
            vsapi->freeFrame(clip);
            vsapi->freeFrame(ref);
            vsapi->setFilterError("vsletterbox: Both inputs must have the same height and width", frameCtx);
            return nullptr;
        }

        auto dst   = vsapi->copyFrame(clip, core);
        auto props = vsapi->getFramePropertiesRW(dst);

        const T * VS_RESTRICT ori_srcp   = reinterpret_cast<const T *>(vsapi->getReadPtr(clip, 0));
        const auto            src_stride = vsapi->getStride(clip, 0) / sizeof(T);
        const T * VS_RESTRICT ori_refp   = reinterpret_cast<const T *>(vsapi->getReadPtr(ref, 0));
        const auto            ref_stride = vsapi->getStride(ref, 0) / sizeof(T);

        int start_y = 0;
        int end_y = height - 1;

        auto srcp   = ori_srcp;
        auto refp   = ori_refp;
        auto stats  = ExponentiallyWeightedStats<>();
        int  bord_y = height;
        auto cutoff = false;
        for (; start_y < height; start_y++) {
            const auto src_mean = calc_15_power_mean<T>(srcp, width);

            const auto st_mean = stats.mean();
            const auto st_stddev = stats.stddev();
            if (st_mean && st_stddev &&
                src_mean > *st_mean + 3 * std::max(*st_stddev, 0.005))
                bord_y = std::min(bord_y, start_y);
            else
                bord_y = height;
            stats.add_data(src_mean);

            if ((cutoff = src_mean > d->thr))
                break;

            srcp += src_stride;
            refp += ref_stride;
        }

        if (!cutoff) {
            start_y = 0;
            end_y = height - 1;
        }
        else [[likely]] {
            if (calc_mean<T>(refp, width) < d->ref_thr)
                start_y = 0;
            if (bord_y < start_y)
                start_y = bord_y;

            srcp   = ori_srcp + end_y * src_stride;
            refp   = ori_refp + end_y * ref_stride;
            stats  = ExponentiallyWeightedStats<>();
            bord_y = -1;
            cutoff = false;
            for (; end_y >= 0; end_y--) {
                const auto src_mean = calc_15_power_mean<T>(srcp, width);

                const auto st_mean = stats.mean();
                const auto st_stddev = stats.stddev();
                if (st_mean && st_stddev &&
                    src_mean > *st_mean + 3 * std::max(*st_stddev, 0.005))
                    bord_y = std::max(bord_y, end_y);
                else
                    bord_y = -1;
                stats.add_data(src_mean);

                if ((cutoff = src_mean > d->thr))
                    break;

                srcp -= src_stride;
                refp -= ref_stride;
            }
            if (!cutoff) {
                vsapi->freeFrame(clip);
                vsapi->freeFrame(ref);
                vsapi->freeFrame(dst);
                vsapi->setFilterError("vsletterbox: Unexpected internal bug \"exceed\"", frameCtx);
                return nullptr;
            }
            else [[likely]] {
                if (calc_mean<T>(refp, width) < d->ref_thr)
                    end_y = height - 1;
                if (bord_y > end_y)
                    end_y = bord_y;
            }
        }

        const auto start_y_prop = "VSLETTERBOX_TOP_ROW";
        const auto end_y_prop   = "VSLETTERBOX_BOTTOM_ROW";
        vsapi->mapSetInt(props, start_y_prop, start_y, maReplace);
        vsapi->mapSetInt(props, end_y_prop, end_y, maReplace);
            
        vsapi->freeFrame(clip);
        vsapi->freeFrame(ref);
    
        return dst;
    }

    return nullptr;
}

static void VS_CC letterbox_search_free(void *instanceData, VSCore *core, const VSAPI *vsapi) {
    auto *d = static_cast<FindData *>(instanceData);
    vsapi->freeNode(d->clip);
    vsapi->freeNode(d->ref);
    delete d;
}

static void VS_CC letterbox_search_create(const VSMap *in, VSMap *out, void *userData, VSCore *core, const VSAPI *vsapi) {
    std::unique_ptr<FindData> d(new FindData);

    d->clip           = vsapi->mapGetNode(in, "clip", 0, nullptr);
    const auto vi     = vsapi->getVideoInfo(d->clip);
    d->ref            = vsapi->mapGetNode(in, "ref", 0, nullptr);
    const auto ref_vi = vsapi->getVideoInfo(d->ref);
    if (!vsh::isConstantVideoFormat(vi) || !vsh::isConstantVideoFormat(ref_vi) ||
        vi->format.sampleType != ref_vi->format.sampleType ||
        vi->format.bitsPerSample != ref_vi->format.bitsPerSample ||
        vi->height != ref_vi->height ||
        vi->width != ref_vi->width) {
        vsapi->mapSetError(out, "vsletterbox: Both input must be of constant format, and of the same sample type, bits per sample, height, and width");
        vsapi->freeNode(d->clip);
        vsapi->freeNode(d->ref);
        return;
    }

    d->thr     = vsapi->mapGetFloat(in, "thr", 0, nullptr);
    d->ref_thr = vsapi->mapGetFloat(in, "ref_thr", 0, nullptr);
    
    VSFilterDependency deps[] = {{d->clip, rpStrictSpatial}, {d->ref, rpStrictSpatial}};
    int num_deps = 2;

    if (vi->format.sampleType == stInteger && vi->format.bitsPerSample == 32)
        vsapi->createVideoFilter(out, "Find", vi, letterbox_search_get_frame<uint32_t>, letterbox_search_free, fmParallel, deps, num_deps, d.release(), core);
    else if (vi->format.sampleType == stInteger && vi->format.bitsPerSample == 16)
        vsapi->createVideoFilter(out, "Find", vi, letterbox_search_get_frame<uint16_t>, letterbox_search_free, fmParallel, deps, num_deps, d.release(), core);
    else if (vi->format.sampleType == stInteger && vi->format.bitsPerSample == 8)
        vsapi->createVideoFilter(out, "Find", vi, letterbox_search_get_frame<uint8_t>, letterbox_search_free, fmParallel, deps, num_deps, d.release(), core);
    else if (vi->format.sampleType == stFloat && vi->format.bitsPerSample == 32)
        vsapi->createVideoFilter(out, "Find", vi, letterbox_search_get_frame<float>, letterbox_search_free, fmParallel, deps, num_deps, d.release(), core);
    else {
        vsapi->mapSetError(out, "vsletterbox: Only 32-bit, 16-bit, or 8-bit integer format or 32-bit float format are supported");
        vsapi->freeNode(d->clip);
        vsapi->freeNode(d->ref);
        return;
    }
}

VS_EXTERNAL_API(void) VapourSynthPluginInit2(VSPlugin *plugin, const VSPLUGINAPI *vspapi) {
    vspapi->configPlugin("aka.letterbox", "letterbox", "Letterbox Detection and Cleaning", VS_MAKE_VERSION(1, 0), VAPOURSYNTH_API_VERSION, 0, plugin);
    vspapi->registerFunction("Find", "clip:vnode;"
                                     "thr:float;"
                                     "ref:vnode;"
                                     "ref_thr:float;", "clip:vnode;", letterbox_search_create, NULL, plugin);
}

File: vsletterbox/vsletterbox.py
# vapoursynth-letterbox
# Copyright (c) Akatsumekusa and contributors

# ---------------------------------------------------------------------
# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the
# "Software"), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject to
# the following conditions:
# 
# The above copyright notice and this permission notice shall be
# included in all copies or substantial portions of the Software.
# 
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
# BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
# ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
# CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.
# ---------------------------------------------------------------------

from vsexprtools import norm_expr
from functools import partial
from vskernels import Bilinear
from vsmasktools import ExKirsch, Morpho
from vstools import ChromaLocation, ColorRange, core, get_y, join, split, vs

def find_letterbox(
        clip,
        permanent=[0, 0],
        dynamic_ref=ExKirsch().edgemask,
        dynamic_ref_thr=2/3
    ):
    assert permanent[0] >= 0
    assert permanent[1] >= 0
    assert permanent[0] + permanent[1] < clip.height - 4
    permanent_top_row = permanent[0]
    permanent_bottom_row = clip.height - 1 - permanent[1]
    assert dynamic_ref_thr >= 0.0 and dynamic_ref_thr <= 1.0

    assert clip.format.num_planes in [1, 3]
    if clip.format.num_planes == 3:
        pre = Bilinear().scale(clip, format=clip.format.replace(subsampling_w=0, subsampling_h=0), range=ColorRange.FULL)
        pre = norm_expr(split(pre), "x y neutral - abs 0.25 * z neutral - abs 0.25 * + + 0 max mask_max min")
    else:
        pre = norm_expr(clip, "x 0 max mask_max min")

    mask = dynamic_ref(pre)
    letterbox = pre.letterbox.Find(ref=mask, ref_thr=dynamic_ref_thr)
    letterbox = letterbox.akarin.PropExpr(lambda: dict(
        VSLETTERBOX_TOP_ROW=f"x.VSLETTERBOX_TOP_ROW {permanent_top_row} 1 + <= {permanent_top_row} x.VSLETTERBOX_TOP_ROW ?",
        VSLETTERBOX_BOTTOM_ROW=f"x.VSLETTERBOX_BOTTOM_ROW {permanent_bottom_row} 1 - >= {permanent_bottom_row} x.VSLETTERBOX_BOTTOM_ROW ?"
    ))
    return letterbox

def letterbox_mask(
        clip,
        permanent=[0, 0],
        dynamic_ref=ExKirsch().edgemask,
        dynamic_ref_thr=2/3,
        fullblack_thr=1/5
    ):
    assert permanent[0] >= 0
    assert permanent[1] >= 0
    assert permanent[0] + permanent[1] < clip.height - 4
    permanent_top_row = permanent[0]
    permanent_bottom_row = clip.height - 1 - permanent[1]
    assert dynamic_ref_thr >= 0.0 and dynamic_ref_thr <= 1.0
    assert fullblack_thr >= 0.0 and fullblack_thr <= 1.0

    letterbox = find_letterbox(clip, permanent, dynamic_ref, dynamic_ref_thr)
    letterbox = letterbox.akarin.PropExpr(lambda: dict(_VSLETTERBOX_AREA_MULTIPLIER=f"""
x.VSLETTERBOX_BOTTOM_ROW 1 + x.VSLETTERBOX_TOP_ROW - {permanent_bottom_row} 1 + {permanent_top_row} - /
{fullblack_thr} /
"""))
    letterbox_mask = norm_expr(letterbox, f"""
Y {permanent_top_row} < mask_max
    Y x.VSLETTERBOX_TOP_ROW < mask_max x._VSLETTERBOX_AREA_MULTIPLIER *
        Y {permanent_bottom_row} > mask_max
            Y x.VSLETTERBOX_BOTTOM_ROW > mask_max x._VSLETTERBOX_AREA_MULTIPLIER *
            0 ? ? ? ?
""")
    return letterbox_mask

def clean_letterbox(
        clip,
        thr=0.030,
        transition=0.020,
        permanent=[0, 0],
        dynamic=True,
        dynamic_ref=ExKirsch().edgemask,
        dynamic_ref_thr=2/3,
        fullblack_thr=1/5,
        border_y=None,
        border_u=None,
        border_v=None
    ):
    assert permanent[0] >= 0
    assert permanent[1] >= 0
    assert permanent[0] + permanent[1] < clip.height - 4
    permanent_top_row = permanent[0]
    permanent_bottom_row = clip.height - 1 - permanent[1]
    assert dynamic_ref_thr >= 0.0 and dynamic_ref_thr <= 1.0
    assert fullblack_thr >= 0.0 and fullblack_thr <= 1.0

    if dynamic:
        letterbox = find_letterbox(clip, permanent, dynamic_ref, dynamic_ref_thr)
    else:
        letterbox = get_y(clip)
        letterbox = letterbox.std.SetFrameProps(VSLETTERBOX_TOP_ROW=permanent_top_row, VSLETTERBOX_BOTTOM_ROW=permanent_bottom_row)
    letterbox = letterbox.akarin.PropExpr(lambda: dict(_VSLETTERBOX_AREA_MULTIPLIER=f"""
x.VSLETTERBOX_BOTTOM_ROW 1 + x.VSLETTERBOX_TOP_ROW - {permanent_bottom_row} 1 + {permanent_top_row} - /
{fullblack_thr} /
"""))
    letterbox_mask = norm_expr(letterbox, f"""
plane_max plane_min - range!
{thr} range@ * plane_min + thr!
{transition} range@ * transition!
Y {permanent_top_row} < mask_max thr@ x - transition@ / 0 1 clamp *
    Y x.VSLETTERBOX_TOP_ROW < mask_max x._VSLETTERBOX_AREA_MULTIPLIER * thr@ x - transition@ / 0 1 clamp *
        Y {permanent_bottom_row} > mask_max thr@ x - transition@ / 0 1 clamp *
            Y x.VSLETTERBOX_BOTTOM_ROW > mask_max x._VSLETTERBOX_AREA_MULTIPLIER * thr@ x - transition@ / 0 1 clamp *
            0 ? ? ? ?
""")
    letterbox_mask = Morpho.minimum(letterbox_mask)

    assert clip.format.num_planes in [1, 3]
    if clip.format.num_planes == 3:
        if clip.format.subsampling_h or clip.format.subsampling_w:
            letterbox_mask_uv = Morpho.minimum(letterbox_mask)
            letterbox_mask = join(letterbox_mask, letterbox_mask_uv, letterbox_mask_uv)
            letterbox_mask = Bilinear().scale(letterbox_mask, format=clip.format, chromaloc=ChromaLocation.from_video(clip))
        else:
            letterbox_mask = join(letterbox_mask, letterbox_mask, letterbox_mask)
    clean = norm_expr([clip, letterbox_mask], ("""
y mask_max / mask!
x plane_min - 1 mask@ - * plane_min +
""", """
y mask_max / mask!
x 1 mask@ - * neutral mask@ * +
"""))
    clean = core.akarin.PropExpr([clean, letterbox_mask], lambda: dict(
        VSLETTERBOX_TOP_ROW="y.VSLETTERBOX_TOP_ROW",
        VSLETTERBOX_BOTTOM_ROW="y.VSLETTERBOX_BOTTOM_ROW",
        _VSLETTERBOX_AREA_MULTIPLIER="y._VSLETTERBOX_AREA_MULTIPLIER"
    ))

    if border_y or border_u or border_v:
        def border(clip, top_row, bottom_row, f):
            combine = []
            if top_row != 0:
                process = clip.std.CropAbs(top=0, height=top_row, width=clip.width)
                combine.append(process)
            process = clip.std.CropAbs(top=top_row, height=bottom_row+1-top_row, width=clip.width)
            process = f(process)
            combine.append(process)
            if bottom_row != clip.height - 1:
                process = clip.std.CropAbs(top=bottom_row+1, height=clip.height-bottom_row-1, width=clip.width)
                combine.append(process)
            return core.std.StackVertical(combine)

        clean_planes = split(clean)
        for plane in range(clean.format.num_planes):
            plane_name = ["y", "u", "v"][plane]
            border_p = locals()[f"border_{plane_name}"]
            if border_p:
                clean_p = clean_planes[plane]
                if not plane or not clip.format.subsampling_h:
                    standard_border_p = border(clean_p, permanent_top_row, permanent_bottom_row, border_p)
                    def border_eval(n, f, clean_p, standard_border_p, border, border_p):
                        if f.props["VSLETTERBOX_TOP_ROW"] == permanent_top_row and \
                           f.props["VSLETTERBOX_BOTTOM_ROW"] == permanent_bottom_row:
                            return standard_border_p
                        else:
                            return border(clean_p, f.props["VSLETTERBOX_TOP_ROW"], f.props["VSLETTERBOX_BOTTOM_ROW"], border_p)
                    process_p = core.std.FrameEval(clean_p, partial(border_eval, clean_p=clean_p, standard_border_p=standard_border_p, border=border, border_p=border_p), prop_src=clean_p)
                else:
                    assert clip.format.subsampling_h == 1
                    standard_border_p = border(clean_p, permanent_top_row >> 1, (permanent_bottom_row + 1) >> 1, border_p)
                    def border_eval(n, f, clean_p, standard_border_p, border, border_p):
                        if f.props["VSLETTERBOX_TOP_ROW"] == permanent_top_row >> 1 and \
                           f.props["VSLETTERBOX_BOTTOM_ROW"] == (permanent_bottom_row + 1) >> 1:
                            return standard_border_p
                        else:
                            return border(clean_p, f.props["VSLETTERBOX_TOP_ROW"] >> 1, (f.props["VSLETTERBOX_BOTTOM_ROW"] + 1) >> 1, border_p)
                    process_p = core.std.FrameEval(clean_p, partial(border_eval, clean_p=clean_p, standard_border_p=standard_border_p, border=border, border_p=border_p), prop_src=clean_p)
                clean_planes[plane] = process_p
        clean = join(clean_planes)

    return clean
